boot/bootstate20.go

```diff
@@ -21,228 +21,176 @@ package boot
-	// we can determine if we are trying the try snap by the presence of
-	// 1. a try-kernel.efi symlink
-	// 2. kernel_status == trying
+// bootState20Kernel implements the bootState and bootStateUpdate interfaces for
+// kernel snaps on UC20. It is used for setNext() and markSuccessful() - though
+// note that for markSuccessful() a different bootStateUpdate implementation is
+// returned, see bootState20MarkSuccessful
+type bootState20Kernel struct {
+	// the bootloader to manipulate kernel assets
+	ebl bootloader.ExtractedRunKernelImageBootloader
 
+	// the kernel_status variable currently in the bootloader, initialized by
+	// setupBootloader()
+	kernelStatus string
+
+	// the kernel_status value to be written in commit()
+	commitKernelStatus string
+
+	// the kernel snap that was tried for markSuccessful()
+	triedKernelSnap snap.PlaceInfo
+
+	// the kernel snap to try for setNext()
+	tryKernelSnap snap.PlaceInfo
+
+	// a flag for revisions() to inform markSuccessful() to clean up status var
+	// if it is inconsistent with what's on disk/extracted
+	markSuccessCleanStatus bool
+}
+
+func (ks20 *bootState20Kernel) loadBootloader() error {
+	// don't setup multiple times
+	if ks20.ebl != nil {
+		return nil
+	}
+	// find the bootloader and ensure it's an extracted run kernel image
+	// bootloader
 	bl, err := bootloader.Find("", nil)
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot find any bootloader: %v", err)
+		return err
 	}
-
 	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
 	if !ok {
-		return nil, nil, false, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", ebl.Name())
-	}
-
-	// get the kernel for this bootloader
-	bootSnap, err = ebl.Kernel()
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ebl.Name(), err)
+		return fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", bl.Name())
 	}
 
-	tryKernel, tryKernelExists, err := ebl.TryKernel()
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ebl.Name(), err)
-	}
+	ks20.ebl = ebl
 
-	if tryKernelExists {
-		tryBootSnap = tryKernel
-	}
-
-	m, err := bl.GetBootVars("kernel_status")
+	// also get the kernel_status
+	m, err := ebl.GetBootVars("kernel_status")
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ebl.Name(), err)
+		return err
 	}
 
-	trying = m["kernel_status"] == "trying"
+	ks20.kernelStatus = m["kernel_status"]
 
-	return bootSnap, tryBootSnap, trying, nil
+	return nil
 }
 
-func (s20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
-	var u20 *bootStateUpdate20
-	var err error
-	u20, err = newBootStateUpdate20(update, snap.TypeKernel)
+func (ks20 *bootState20Kernel) revisions() (snap.PlaceInfo, snap.PlaceInfo, bool, error) {
```

Comment:
my suggestion was to turn the bool return value, into returning  status string

yes, this requires a small change in boot.go and bootstate16.go as well, can be done here or as a small separate PR